### ISSUE TITLE
Checking for member of TypedDict does not narrow type until later branch

### ISSUE THREAD

**Issue by SebbyLaw** [astral-sh/ty#2655]
### Summary

Not sure if this is a duplicate of #1479 
https://play.ty.dev/1adac9d8-0953-4ed0-b596-7400450e5db9
```py
from typing import TypedDict, reveal_type


class Foo(TypedDict):
    foo: str
    value: int


class Bar(TypedDict):
    bar: str
    data: float


def process_foobar(item: Foo | Bar) -> str:
    if "foo" in item:
        reveal_type(item)  # Revealed type: `Foo | Bar`
        return item["foo"]  # error
    else:
        # correctly narrows to Bar?
        reveal_type(item)  # Revealed type: `Bar`
        return item["bar"]
```

### Version

ty 0.0.14 (16597f5e2 2026-01-26)

**Comment by AlexWaygood** [comment]
Thanks for the report! However, this is intentional. We cannot say that `item` is definitely not an instance of `Bar` just because the `"foo"` key exists in `item`. It's possible to subclass `Bar` and add a `"foo"` key:

```py
from typing import TypedDict, reveal_type


class Foo(TypedDict):
    foo: str
    value: int


class Bar(TypedDict):
    bar: str
    data: float


class BarSub(Bar):
    foo: str


def process_foobar(item: Foo | Bar) -> str:
    if "foo" in item:
        reveal_type(item)  # Revealed type: `Foo | Bar`
        return item["foo"]  # error
    else:
        # correctly narrows to Bar?
        reveal_type(item)  # Revealed type: `Bar`
        return item["bar"]


process_foobar(item=BarSub({"bar": "", "data": 3.14, "foo": ""})
```

The final `process_foobar()` call is accepted by mypy, ty, pyright and pyrefly, which is why it's unsound to narrow `item` to `Foo` in the `if "foo" in item` branch.

Mypy has the same behaviour here as ty here. Pyright does appear to narrow the type to `Foo` in the first branch; but pyright's behaviour, I would argue, is unsound in this case. Pyrefly doesn't seem to narrow the type here at all yet.

**Comment by AlexWaygood** [comment]
The reason why we can nonetheless narrow the type in the `else` branch here is that it's not possible for `TypedDict` subclasses to _remove_ keys, only add them. So if `"foo"` is _not_ in `item`, that definitively rules out the possibility that `item` could be an instance of `Foo`; it can only then be an instance of `Bar`.

**Comment by SebbyLaw** [comment]
I see. Would the recommended solution here be using the PEP728 `closed=True` class parameter? (and waiting for support from ty https://github.com/astral-sh/ty/issues/154)

**Comment by AlexWaygood** [comment]
yes, marking these `TypedDict`s as `closed=True` would allow us to narrow in the first branch (once we've implemented support for PEP728)
